File: src/routes/cashRegister.js
const express = require('express');
const cashRegisterService = require('../services/cashRegister');
const { ValidationError } = require('../services/errors');

const router = express.Router();

function handleError(res, err) {
  if (err instanceof ValidationError) {
    return res.status(400).json({ error: err.message });
  }
  console.error(err);
  return res.status(500).json({ error: 'Erro interno do servidor.' });
}

router.get('/current', (req, res) => {
  res.json(cashRegisterService.currentSummary());
});

router.post('/open', (req, res) => {
  try {
    const register = cashRegisterService.open(req.body.openingAmount);
    res.status(201).json(register);
  } catch (err) {
    handleError(res, err);
  }
});

router.post('/movement', (req, res) => {
  try {
    const { type, amount, reason } = req.body;
    const summary = cashRegisterService.registerMovement(type, amount, reason);
    res.status(201).json(summary);
  } catch (err) {
    handleError(res, err);
  }
});

router.post('/close', (req, res) => {
  try {
    const result = cashRegisterService.close(req.body.countedAmount);
    res.json(result);
  } catch (err) {
    handleError(res, err);
  }
});

router.get('/closures', (req, res) => {
  res.json(cashRegisterService.listClosures());
});

module.exports = router;

File: internal/httpapi/cashregister.go
package httpapi

import (
	"encoding/json"
	"net/http"
	"strings"
)

func (h *Handlers) getCurrentRegister(w http.ResponseWriter, r *http.Request) {
	summary, err := h.cashRegister.CurrentSummary()
	if err != nil {
		writeError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, summary)
}

func (h *Handlers) openRegister(w http.ResponseWriter, r *http.Request) {
	var body struct {
		OpeningAmount flexNumber `json:"openingAmount"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)

	register, err := h.cashRegister.Open(body.OpeningAmount.Value, body.OpeningAmount.Present && body.OpeningAmount.Valid)
	if err != nil {
		writeError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, register)
}

func (h *Handlers) registerCashMovement(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Type   string     `json:"type"`
		Amount flexNumber `json:"amount"`
		Reason string     `json:"reason"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)

	summary, err := h.cashRegister.RegisterMovement(
		body.Type, body.Amount.Value, body.Amount.Present && body.Amount.Valid, strings.TrimSpace(body.Reason),
	)
	if err != nil {
		writeError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, summary)
}

func (h *Handlers) closeRegister(w http.ResponseWriter, r *http.Request) {
	var body struct {
		CountedAmount flexNumber `json:"countedAmount"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)

	result, err := h.cashRegister.Close(body.CountedAmount.Value, body.CountedAmount.Present && body.CountedAmount.Valid)
	if err != nil {
		writeError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *Handlers) listClosures(w http.ResponseWriter, r *http.Request) {
	closures, err := h.cashRegister.ListClosures()
	if err != nil {
		writeError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, closures)
}
